equihash/equihash.go

```diff
@@ -0,0 +1,723 @@
+package consensus
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"hash"
+	"math"
+	"math/big"
+	"reflect"
+	"sort"
+
+	blake2b "github.com/minio/blake2b-simd"
+)
+
+const (
+	// size of word
+	wordSize = 32
+	wordMask = (1 << wordSize) - 1
+	byteMask = 0xFF
+	// size of each hash key
+	hashSize = 64
+	// N is the number of hash digests used to find a mining solution
+	N = 96
+	// K is the exponent for xor'ing 2^k hash digests for solution
+	K = 5
+	// person prefix for excc
+	exccPrefix = "excc"
+)
+
+var (
+	errBadArg           = errors.New("invalid argument")
+	errWriteLen         = errors.New("didn't write full len")
+	errKLarge           = errors.New("k should be less than n")
+	errCollisionLen     = errors.New("collision length too big")
+	errSmallBitLen      = errors.New("bitLen < 8")
+	errSmallWordSize    = errors.New("wordSize < 7+bitLen")
+	errBadOutLen        = errors.New("outLen != 8*outWidth*len(in)/bitLen")
+	errDuplicateIndices = errors.New("duplicate indices")
+	errPairWiseOrdering = errors.New("bad pair-wise ordering")
+	errBadWord          = errors.New("bad word")
+	bigZero             = big.NewInt(0)
+)
+
+// the generic person prefix encoder; which encodes the prefix and gbp parameters
+func person(prefix string, n, k int) []byte {
+	nb, kb := writeUint32(uint32(n)), writeUint32(uint32(k))
+	return append([]byte(prefix), append(nb, kb...)...)
+}
+
+// returns the encoded excc person
+func exccPerson(n, k int) []byte {
```

Comment:
Yeah, looks very much redundant after I made prefix as a parameter to `gbp`. It can be removed and its underlying tests iff (if and only if) we will never use `const prefix = "excc"` but if we are going to use that prefix then it's still important to test it; so would move it to the test file and it's an additional `person` test. Thoughts?

Reply:
I'm not sure if we will need it. @siy can you help?